collector/filesystem_freebsd.go

```diff
@@ -21,78 +21,65 @@ import (
 import "C"
 
 const (
-	filesystemSubsystem = "filesystem"
+	subsystem = "filesystem"
 )
 
 var (
-	ignoredMountPoints = flag.String("collector.filesystem.ignored-mount-points", "^/(dev)($|/)", "Regexp of mount points to ignore for filesystem collector.")
+	ignoredMountPoints = flag.String(
+		"collector.filesystem.ignored-mount-points",
+		"^/(dev)($|/)",
+		"Regexp of mount points to ignore for filesystem collector.")
 )
 
 type filesystemCollector struct {
 	ignoredMountPointsPattern *regexp.Regexp
-
-	size, free, avail, files, filesFree *prometheus.GaugeVec
 }
 
 func init() {
 	Factories["filesystem"] = NewFilesystemCollector
 }
 
-// Takes a prometheus registry and returns a new Collector exposing
-// Filesystems stats.
 func NewFilesystemCollector() (Collector, error) {
-	var filesystemLabelNames = []string{"filesystem"}
-
+	pattern := regexp.MustCompile(*ignoredMountPoints)
 	return &filesystemCollector{
-		ignoredMountPointsPattern: regexp.MustCompile(*ignoredMountPoints),
-		size: prometheus.NewGaugeVec(
-			prometheus.GaugeOpts{
-				Namespace: Namespace,
-				Subsystem: filesystemSubsystem,
-				Name:      "size_bytes",
-				Help:      "Filesystem size in bytes.",
-			},
-			filesystemLabelNames,
-		),
-		free: prometheus.NewGaugeVec(
-			prometheus.GaugeOpts{
-				Namespace: Namespace,
-				Subsystem: filesystemSubsystem,
-				Name:      "free_bytes",
-				Help:      "Filesystem free space in bytes.",
-			},
-			filesystemLabelNames,
-		),
-		avail: prometheus.NewGaugeVec(
-			prometheus.GaugeOpts{
-				Namespace: Namespace,
-				Subsystem: filesystemSubsystem,
-				Name:      "avail_bytes",
-				Help:      "Filesystem space available to non-root users in bytes.",
-			},
-			filesystemLabelNames,
-		),
-		files: prometheus.NewGaugeVec(
-			prometheus.GaugeOpts{
-				Namespace: Namespace,
-				Subsystem: filesystemSubsystem,
-				Name:      "file_nodes",
-				Help:      "Filesystem total file nodes.",
-			},
-			filesystemLabelNames,
-		),
-		filesFree: prometheus.NewGaugeVec(
-			prometheus.GaugeOpts{
-				Namespace: Namespace,
-				Subsystem: filesystemSubsystem,
-				Name:      "file_free_nodes",
-				Help:      "Filesystem total free file nodes.",
-			},
-			filesystemLabelNames,
-		),
+		ignoredMountPointsPattern: pattern,
 	}, nil
 }
 
+var (
+	filesystemLabelNames = []string{"device", "mountpoint", "fstype"}
+
+	sizeDesc = prometheus.NewDesc(
```

Comment:
Since all collectors are in the same package, the naming here is polluting the namespace. Many collectors might need a `sizeDesc`. You could either prefix these names with `filesystem...`, or put them into the filesystemCollector struct (where you could also put the `subsystem`). Fundamentally, we should probably have each collector in their own package, or at least the more complex ones... but that's a different discussion, which we should not drag into this PR.